src/cloud-api-adaptor/pkg/securecomms/ppssh/ppssh.go

```diff
@@ -0,0 +1,304 @@
+		}
+		//ctx, cancel := context.WithCancel(context.Background())
+		//defer cancel()
+
+		log.Printf("Kubernetes client connected\n")
+		peer, err = KubernetesSShService(ctx, nConn, kubernetesPhaseConfig)
+		if err != nil {
+			log.Printf("Failed k8sPhase: %s", err)
+			peer = nil
+			continue
+		}
+		peer.AddOutbounds(outbounds)
+		err = peer.AddInbounds(inbounds)
+		if err != nil {
+			log.Printf("Failed addInbounds during k8sPhase: %s", err)
+			return
+		}
+		peer.Wait()
+		log.Printf("KubernetesSShService exiting")
+	}
+}
+
+func attestationPhase(listener net.Listener, inbounds sshproxy.Inbounds, outbounds sshproxy.Outbounds, ppSecrets *PpSecrets) *ssh.ServerConfig {
+	// Singleton - accept an unproven connection for attestation
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	var peer *sshproxy.SshPeer
+	for peer == nil {
+		log.Printf("Attastation Phase: waiting for client to connect\n")
+		nConn, err := listener.Accept()
+		if err != nil {
+			log.Fatal("Attastation Phase: failed to accept incoming connection: ", err)
+		}
+
+		log.Printf("Attastation Phase: client connected\n")
+
+		peer, err = AttestationSShService(ctx, nConn)
+		if err != nil {
+			log.Print(err.Error())
+			peer = nil
+		}
+	}
+
+	peer.AddOutbounds(outbounds)
+	if err := peer.AddInbounds(inbounds); err != nil {
+		log.Fatal("Attastation Phase: failed to add Inbounds: ", err)
+	}
+
+	for {
+		log.Printf("Attastation Phase: getting keys from KBS\n")
+		ppSecrets.AddKey(WN_PUBLIC_KEY)
+		ppSecrets.AddKey(PP_PRIVATE_KEY)
+		ppSecrets.Go() // wait for the keys
+		config, err := InitKubernetesPhaseSshConfig(ppSecrets)
+		if err == nil {
+			log.Printf("Attastation Phase: InitKubernetesPhaseSshConfig is ready\n")
+			peer.Upgrade()
+			return config
+		}
+		log.Printf("Attastation Phase: failed getting keys from KBS: %v\n", err)
+	}
+	// wait for a provenPpPrivateKey
+	//WaitForProvenKeys(ctx, peer)
+	//peer.Wait()
+}
+
+func InitSshServer(inbound_strings, outbounds_strings []string, getSecret GetSecret) {
+	log.Printf("Using PP Secure Comms: InitSshServer version %s", sshutil.PpSecureCommsVersion)
+
+	Singleton()
+	var inbounds sshproxy.Inbounds
+	var outbounds sshproxy.Outbounds
+	var wg sync.WaitGroup
+	for _, tag := range inbound_strings {
+		_, _, err := inbounds.Add(tag, &wg)
+		if err != nil {
+			log.Fatalf("failed to add inbound %s: %v", tag, err)
+		}
+	}
+	for _, tag := range outbounds_strings {
+		outbounds.Add(tag)
+	}
+
+	log.Printf("SSH Service starting 0.0.0.0:%s\n", sshutil.SSHPORT)
+	listener, err := net.Listen("tcp", "0.0.0.0:"+sshutil.SSHPORT)
+	if err != nil {
+		log.Fatal("failed to listen for connection: ", err)
+	}
+
+	go func() {
+		kubernetesPhaseConfig := attestationPhase(listener, inbounds, outbounds, NewPpSecrets(getSecret))
+		for {
+			k8sPhase(listener, inbounds, outbounds, kubernetesPhaseConfig)
+		}
+	}()
+}
+
+func Singleton() {
+	// Signleton- make sure we run the ssh service once per boot.
```

Comment:
```suggestion
	// Singleton- make sure we run the ssh service once per boot.
```

Reply:
You are looking an an old commit - was already fixed